java-checks/src/main/java/org/sonar/java/checks/PrintfMisuseCheck.java

```diff
@@ -0,0 +1,202 @@
+    }
+    cleanupLineSeparator(params);
+    if (!params.isEmpty() && argIndexes(params).size() <= args.size()) {
+      verifyParameters(mit, args, params);
+    }
+  }
+
+  private void verifyParameters(MethodInvocationTree mit, List<ExpressionTree> args, List<String> params) {
+    int index = 0;
+    List<ExpressionTree> unusedArgs = new ArrayList<>(args);
+    for (String rawParam : params) {
+      String param = rawParam;
+      int argIndex = index;
+      if (param.contains("$")) {
+        argIndex = getIndex(param) - 1;
+        if (argIndex == -1) {
+          return;
+        }
+        param = param.substring(param.indexOf('$') + 1);
+      } else if (param.charAt(0) == '<') {
+        //refers to previous argument
+        argIndex = Math.max(0, argIndex - 1);
+      }else {
+        index++;
+      }
+      ExpressionTree argExpressionTree = args.get(argIndex);
+      unusedArgs.remove(argExpressionTree);
+      Type argType = argExpressionTree.symbolType();
+      checkBoolean(mit, param, argType);
+    }
+    reportUnusedArgs(mit, args, unusedArgs);
+  }
+
+  protected void handleMessageFormat(MethodInvocationTree mit, String formatString, List<ExpressionTree> args) {
+    String newFormatString = cleanupDoubleQuote(formatString);
+    Set<Integer> indexes = getMessageFormatIndexes(newFormatString);
+    List<ExpressionTree> newArgs = args;
+    if (newArgs.size() == 1) {
+      ExpressionTree firstArg = newArgs.get(0);
+      if (firstArg.symbolType().isArray()) {
+        if (isNewArrayWithInitializers(firstArg)) {
+          newArgs = ((NewArrayTree) firstArg).initializers();
+        } else {
+          // size is unknown
+          return;
+        }
+      }
+    }
+    if (checkEmptyParams(mit, indexes)) {
+      return;
+    }
+    checkToStringInvocation(newArgs);
+    verifyParameters(mit, newArgs, indexes);
+  }
+
+  private boolean checkEmptyParams(MethodInvocationTree mit, Collection<?> params) {
+    return checkAndReport(mit, params.isEmpty(), "String contains no format specifiers.");
+  }
+
+  private void checkToStringInvocation(List<ExpressionTree> args) {
+    args.stream()
+      .filter(arg -> arg.is(Tree.Kind.METHOD_INVOCATION) && TO_STRING.matches((MethodInvocationTree) arg))
+      .forEach(arg -> reportIssue(arg, "No need to call toString \"method()\" as formatting and string conversion is done by the Formatter."));
+  }
+
+
+  private void verifyParameters(MethodInvocationTree mit, List<ExpressionTree> args, Set<Integer> indexes) {
+    List<ExpressionTree> unusedArgs = new ArrayList<>(args);
+    for (int index : indexes) {
+      if (index >= args.size()) {
+        return;
+      }
+      unusedArgs.remove(args.get(index));
+    }
+    reportUnusedArgs(mit, args, unusedArgs);
+  }
+
+  private void reportUnusedArgs(MethodInvocationTree mit, List<ExpressionTree> args, List<ExpressionTree> unusedArgs) {
+    for (ExpressionTree unusedArg : unusedArgs) {
+      int i = args.indexOf(unusedArg);
```

Comment:
Not related to the split, although this loop should be done with classic for-loop, if index is needed.